sparse_ho/criterion.py

```diff
@@ -201,76 +179,58 @@ class HeldOutSmoothedHinge():
     TODO
     """
 
-    def __init__(self, X_val, y_val, model, X_test=None, y_test=None):
+    def __init__(self, idx_train, idx_val, X_test=None, y_test=None):
         """
-        Parameters
+        Parameters:
         ----------
-        X_val : {ndarray, sparse matrix} of shape (n_samples, n_features)
-            Validation data
-        y_val : {ndarray, sparse matrix} of shape (n_samples)
-            Validation target
-        model: instance of Model
-            Object of the class Model (e.g. Lasso or Sparse logistic regression)
+        idx_train: np.array
+            indices of the training set
+        idx_test: np.array
+            indices of the testing set
         X_test : {ndarray, sparse matrix} of shape (n_samples_test, n_features)
             Test data
         y_test : {ndarray, sparse matrix} of shape (n_samples_test,)
             Test target
         """
-        self.X_val = X_val
-        self.y_val = y_val
+        self.idx_train = idx_train
+        self.idx_val = idx_val
         self.X_test = X_test
         self.y_test = y_test
-        self.model = model
 
         self.mask0 = None
         self.dense0 = None
         self.quantity_to_warm_start = None
         self.val_test = None
         self.rmse = None
 
-    def get_v(self, mask, dense):
-        Xbeta_y = self.y_val * (self.X_val[:, mask] @ dense)
-        deriv = derivative_smooth_hinge(Xbeta_y)
-        if issparse(self.X_val):
-            v = self.X_val[:, mask].T.multiply(deriv * self.y_val)
-            v = np.array(np.sum(v, axis=1))
-            v = np.squeeze(v)
-        else:
-            v = (deriv * self.y_val)[:, np.newaxis] * self.X_val[:, mask]
-            v = np.sum(v, axis=0)
-        v /= self.X_val.shape[0]
-        return v
-
-    def value(self, mask, dense):
-        if issparse(self.X_val):
-            Xbeta_y = (self.X_val[:, mask].T).multiply(self.y_val).T @ dense
-        else:
-            Xbeta_y = self.y_val * (self.X_val[:, mask] @ dense)
-        return np.sum(smooth_hinge(Xbeta_y)) / self.X_val.shape[0]
-
-    def value_test(self, mask, dense):
-        if self.X_test is not None and self.y_test is not None:
-            if issparse(self.X_test):
-                Xbeta_y = (self.X_test[:, mask].T).multiply(self.y_test).T @ dense
-            else:
-                Xbeta_y = self.y_test * (self.X_test[:, mask] @ dense)
-            self.val_test = np.sum(smooth_hinge(Xbeta_y)) / self.X_test.shape[0]
-        else:
-            self.val_test = None
+    def value_outer_crit(self, X, y, mask, dense):
+        if X is None or y is None:
+            return None
 
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
+        if issparse(X):
+            Xbeta_y = (X[:, mask].T).multiply(y).T @ dense
```

Comment:
```suggestion
            Xbeta_y = (X[:, mask].T @ y).T @ dense
```

Reply:
Actually it's to perform row-wise multiplication while preserving sparsit of the output, it's not vector matrix multiplication